src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+        /// list for the purposes of identifying relevant places for statement completion:</para>
+        /// <list type="bullet">
+        /// <item><description>The closing delimiter is typically inserted by the Automatic Brace Compeltion feature.</description></item>
+        /// <item><description>It is not syntactically valid to place a semicolon <em>directly</em> within the delimiters.</description></item>
+        /// </list>
+        /// </remarks>
+        /// <param name="currentNode"></param>
+        /// <returns><see langword="true"/> if <paramref name="currentNode"/> is a statement that ends with a closing
+        /// delimiter, and that closing delimiter exists in the source code; otherwise, <see langword="false"/>.
+        /// </returns>
+        private static bool StatementClosingDelimiterIsMissing(SyntaxNode currentNode)
+        {
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.DoStatement:
+                    var dostatement = (DoStatementSyntax)currentNode;
+                    return dostatement.CloseParenToken.IsMissing;
+                case SyntaxKind.ForStatement:
+                    var forStatement = (ForStatementSyntax)currentNode;
+                    return forStatement.CloseParenToken.IsMissing;
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Determines if a syntax node includes all required closing delimiters.
+        /// </summary>
+        /// <remarks>
+        /// <para>Some syntax nodes, such as parenthesized expressions, require a matching closing delimiter to end the
+        /// syntax node. If this node is omitted from the source code, the parser will automatically insert a zero-width
+        /// "missing" closing delimiter token to produce a valid syntax tree. This method determines if required closing
+        /// delimiters are present in the original source.</para>
+        /// </remarks>
+        /// <param name="currentNode"></param>
+        /// <returns>
+        /// <list type="bullet">
+        /// <item><description><see langword="true"/> if <paramref name="currentNode"/> requires a closing delimiter and the closing delimiter is present in the source (i.e. not missing)</description></item>
+        /// <item><description><see langword="true"/> if <paramref name="currentNode"/> does not require a closing delimiter</description></item>
+        /// <item><description>otherwise, <see langword="false"/>.</description></item>
+        /// </list>
+        /// </returns>
+        private static bool ClosingDelimiterExistsIfNeeded(SyntaxNode currentNode)
+        {
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.ArgumentList:
+                    var argumentList = (ArgumentListSyntax)currentNode;
+                    return !argumentList.CloseParenToken.IsMissing;
```

Comment:
what about attribute-argument-lists.  For that matter, i haven't really seen how attributes are handled (if at all) in this feature.  Do we have tests around them?

Reply:
Attributes are not candidates for completion since I can't think of any scenarios where we'd add a semicolon.  But I have added tests for attributes with and without argument lists.